### PR TITLE
The C++ sources and the gradle build file

### PR DESCRIPTION
**Pull request by benignbala**
This set of commits has the c++ source files that will be built into a library and a cpp.gradle that actually builds the library

It also includes a native binary that uses the library that is built

### CODE REVIEW

**Comment by benignbala** [comment]
Looks like the travis-ci for XCode timed out. Can someone please help me understand why ? Thanks

**Comment by benignbala** [comment]
Noticed this issue on travis-ci https://github.com/travis-ci/travis-ci/issues/6422 . Wondering if it could be related.

**Comment by ysb33r** [comment]
I'm looking into it.

**Comment by benignbala** [comment]
Thanks @ysb33r

**Comment by ysb33r** [comment]
Try to update the plugin version in `build.gradle` to `0.5.4`. If it does not fix the issue.

**Comment by benignbala** [comment]
Thanks, trying it out now.

**Comment by benignbala** [comment]
This time, the xcode tests launched, but timed out at the gitPublishReset step.

**Comment by ysb33r** [comment]
Let's just try the following code in build.gradle as temporary measure - can rebase it out later.

```
travisci  {
  if(System.getenv('TRAVIS_OS_NAME') != 'linux') {
    tasks.each { t ->
      if( t.name.startsWith('gitPublish') {
        t.enabled = false
      }
    }
  }
}
```

If that works, I can look at fixing it in the Gradle Guide plugin.

**Comment by benignbala** [comment]
That worked :) Thanks a lot sir!

File: contents/index.adoc [hunk 1]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+----
+include::{projdir}/gradle/cpp.gradle[]
+----
+<1> {cpp} libraries and binaries are built using the `cpp` plugin.
+<2> Native builds use the `model` block
+<3> Native libraries are specified as a component of type `NativeLibrarySpec` and defined by a name - `greeter` in this case.
+<4> Native binaries are specified as a component of type `NativeBinarySpec` and defined by a name - `main` in this case.
+<5> The `main` binary uses the `greeter` {c++} library and is specified using `cpp.lib`
+
+Now, if we run
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Be imperative. Rather use _Now run_.

File: contents/index.adoc [hunk 2]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+
+[listing]
+----
+$ ./gradlew tasks
+----
+
+we can see that Gradle has added a number of tasks
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Prefer _you should_ over _we can_. (See the Style Guide).

The purpose is to speak to the reader in the second person, not use a collective first person. Also as a write you have no idea whether the reader picked up what you intended, therefore the use of _should_.

File: contents/index.adoc [hunk 3]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+
+[listing]
+----
+greeterSharedLibrary - Assembles shared library 'greeter:sharedLibrary'.
+greeterStaticLibrary - Assembles static library 'greeter:staticLibrary'.
+installMainExecutable - Installs a development image of executable 'main:executable'
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Remove references to `...Executable...`. The focus is on building libraries.

File: contents/index.adoc [hunk 4]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+----
+
+There are two things to note here
+
+. The names `Greeter` and `Main` in the task names are derived from the names of the components defined in `build.gradle`.
+. The `NativeLibrarySpec`, by default, generates both static and shared(dynamic) libraries.  By default, the shared library is used by the binaries that use this library
+
+== Add source code
+
+In the build.gradle file, we have defined two components, `greeter`, which is a library and `main`, which is a binary. So, by default, the sources for these should be present as follows
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Remove _we_. _In `build.gradle`, there are two components. `greeter` is the library you will be building and `main` is a simple executable to consume this library`_
- [ ] Remove _So, by default_ -> _The sources are structured as below:_

File: contents/index.adoc [hunk 5]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+
+. `greeter` library - `src/greeter/cpp` for sources and _non-exported header files_ and `src/greeter/headers` for _exported header files_
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Ordered list indicates precendence. Use  unordered (`*`) instead.

File: contents/index.adoc [hunk 6]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+include::{projdir}/src/greeter/headers/greeter.hpp[]
+----
+
+.src/greeter/cpp/greeter.cpp
+[source,cpp]
+----
+include::{projdir}/src/greeter/cpp/greeter.cpp[]
+----
+
+Now, place a `greeting.cpp` under `src/main/cpp`.
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Drop the comma

File: contents/index.adoc [hunk 7]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+
+.src/main/cpp/greeting.cpp
+[source,cpp]
+----
+include::{projdir}/src/main/cpp/greeting.cpp[]
+----
+
+== Build the project
+
+As mentioned earlier, Gradle adds tasks to build both the static and shared form of the library. To build just the static library, run
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] _Gradle ads tasks to the build for both_

File: contents/index.adoc [hunk 8]
```diff
@@ -28,11 +30,163 @@ BUILD SUCCESSFUL
...
+[listing]
+----
+$ find build/libs/
+build/libs/
+build/libs/greeter
+build/libs/greeter/static
+build/libs/greeter/static/libgreeter.a
+----
+
+In a similar fasion, the shared library can be generated by running
```
Comment (ysb33r) [inline contents/index.adoc]: - [ ] Typo: _In a similar fashion_

**Comment by ysb33r** [comment]
@benignbala I have completed a first round of review and left my comments. Not yet executed your steps, but I'll leave that for when your changes are done.

**Comment by benignbala** [comment]
@ysb33r Thanks for the inputs and feedback. I've amended the style and updated #1

**Comment by ysb33r** [comment]
My suggestion is that we add the testing to this guide using GoogleTest. However this is going to be quite a bit more work as we'll need to ensure GoogleTest is placed into CI automatically.

I am going to merge this as it covers enough detail, besides testing mentioned above. Then we can get someone else to also review the guide as it is now. It also means that we can take it live.

You will then be able to work on the testing part in a feature branch until it is ready.

**Comment by benignbala** [comment]
Thanks @ysb33r. You've been extremely helpful (and patient) in reviewing and guiding me through this :). I'll create a new feature branch and start working  on testing aspects once we've this reviewed by others too. Thanks

**Pull request merged**